Readme.md
# mock-generator [![NPM version][npm-image]][npm-url] [![build status][travis-image]][travis-url] [![Test coverage][coveralls-image]][coveralls-url]

> Random data generator

## Installation

    npm install mock-generator

## License

MIT

[npm-image]: https://img.shields.io/npm/v/mock-generator.svg?style=flat
[npm-url]: https://npmjs.org/package/mock-generator
[travis-image]: https://img.shields.io/travis/meituan/mock-generator.svg?style=flat
[travis-url]: https://travis-ci.org/meituan/mock-generator
[coveralls-image]: https://img.shields.io/coveralls/meituan/mock-generator.svg?style=flat
[coveralls-url]: https://coveralls.io/r/meituan/mock-generator?branch=master


History.md
1.0.1 / 2015-01-28
==================

  * Fix range data mingled with pick


test/generator.js
var random = require('..');
var _ = require('lodash');

function range(input, min, max) {
    return input >= min && input <= max;
}

var rEmail = /[\w.]+@\w+\.\w+/,
    rDate = /\d{4}-\d{2}-\d{2}/,
    rTime = /\d{2}:\d{2}:\d{2}/,
    rDatetime = /\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}/,
    rFloat = /(\d+)\.?(\d+)?/;

describe('mock-random', function() {
    this.timeout(5000);
    it('basic', function() {
        random(1).should.eql(1);

        random(true).should.eql(true);
        random(false).should.eql(false);

        random('').should.eql('');
        random('foo').should.eql('foo');

        random([]).should.eql([]);
        random({}).should.eql({});

        random('@EMAIL').should.not.eql('@EMAIL');
    });

    it('orig array', function() {
        var arr = [{
            foo: 'foo'
        }, {
            bar: 'bar'
        }, {
            foobar: 'foobar'
        }];
        var data = random({
            arr: arr
        });
        data.arr.length.should.eql(3);
        data.arr.should.not.equal(arr);
        for (var i = 0; i < data.arr.length; i++) {
            data.arr[i].should.not.equal(arr[i]);
        }
    });

    it('array', function() {
        var data;

        data = random({
            'list|1': [{}]
        });
        data.list.length.should.eql(1);

        data = random({
            'list|10': [{}]
        });
        data.list.length.should.eql(10);

        data = random({
            'list|5-10': [{}]
        });
        data.list.length.should.above(4);
        data.list.length.should.below(11);
    });

    it('pick', function() {
        function target(name, value) {
            var tpl = {};
            tpl[name] = value;

            var data = random(tpl);
            (typeof data.opt).should.eql(typeof value[0]);
            value.indexOf(data.opt).should.above(-1);
        }

        target('opt|1', [1, 2, 4, 8]);
        target('opt|1', ['GET', 'POST', 'HEAD', 'DELETE']);
    });

    it('range with pick', function() {
        var data = random({
            'list|5-10': [1, 2]
        });
        data.list.length.should.above(4);
        data.list.length.should.below(11);
    });

    it('pick object from array', function() {
        var tpl = {
            'opt|1': [{ // 对备选元素会再次做解析
                method: 'GET'
            }, {
                method: 'POST'
            }, {
                method: 'HEAD'
            }, {
                method: 'DELETE'
            }]
        };
        var data = random(tpl);
        (typeof data.opt).should.eql(typeof tpl['opt|1'][0]);
        for (var i = 0; i < tpl['opt|1'].length; i++) {
            if (data.opt.method === tpl['opt|1'][i].method) {
                return;
            }
        }
        (0).should.be.ok;
    });

    it('distinguish range from pick', function() {
        var data = random({
            'list|1-1': [{}, {}]
        });
        data.list.length.should.eql(1);
    });

    it('pick from object', function() {
        var tpl = {
            'pick1|1': {
                get: '@URL',
                post: '@URL',
                head: '@URL',
                put: '@URL',
                'delete': '@URL'
            }
        };
        tpl['pick2|2'] = tpl['pick1|1'];
        tpl['pick3|3'] = tpl['pick1|1'];
        tpl['pick6|6'] = tpl['pick1|1'];

        var data = random(tpl);
        _.keys(data.pick1).length.should.eql(1);
        _.keys(data.pick2).length.should.eql(2);
        _.keys(data.pick3).length.should.eql(3);
        _.keys(data.pick6).length.should.eql(5);
    });

    it('float', function() {
        function target(tpl, min, max, dmin, dmax) {
            var data, parts, i;
            for (i = 0; i < 1000; i++) {
                data = random(tpl);
                rFloat.lastIndex = 0;
                parts = rFloat.exec('' + data.float); // 可能会自动转为整数，例如 10.0 > 10

                (typeof data.float).should.eql('number');
                (data.float >= min && data.float <= max, JSON.stringify(tpl)).should.be.ok;
                if (parts[2]) {
                    (parts[2].length >= dmin && parts[2].length <= dmax, JSON.stringify(tpl)).should.be.ok;
                }
            }
        }

        target({
            'float|.1-10': 10
        }, 10, 11, 1, 10);

        target({
            'float|.3-10': 123.123
        }, 123, 124, 3, 10);

        target({
            'float|20-100.1-10': 10
        }, 20, 101, 1, 10);

        target({
            'float|99.1-10': 10
        }, 99, 100, 1, 10);

    });

    it('integer', function() {
        function target(tpl, min, max) {
            var data, i;
            for (i = 0; i < 1000; i++) {
                data = random(tpl);
                (typeof data.integer).should.eql('number');
                (data.integer != 1).should.be.ok;
                (data.integer >= min && data.integer <= max).should.be.ok;
            }
        }

        target({
            'integer|2-100': 1
        }, 2, 100);

        target({
            'integer|100-2': 1
        }, 2, 100);

        target({
            'integer|2-2': 1
        }, 2, 2);
    });

    it('string', function() {
        function target(tpl, min, max) {
            var data, i;
            for (i = 0; i < 1000; i++) {
                data = random(tpl);

                (typeof data.string).should.eql('string');
                (data.string.length >= min && data.string.length <= max).should.be.ok;
            }
        }

        target({
            'string|1-10': '★号'
        }, 2, 20);

        target({
            'string|10': '★号'
        }, 20, 20);

    });

    it('boolean', function() {
        var data, count = 0, i;
        for (i = 0; i < 1000; i++) {
            data = random({
                'bool|1': false
            });

            (typeof data.bool === 'boolean').should.be.ok;
            if (data.bool) count++;
        }
        range(count, 300, 700).should.be.ok; // 可能会失败，但是仍在预期范围内，因为结果毕竟是随机的。
    });

    it('function', function() {
        var tpl = {
            prop: 'hello',
            fn: function() {
                return this.prop;
            }
        };
        var data = random(tpl);
        data.fn.should.eql('hello');
    });

    it('disorderly function', function() {
        var tpl = {
            xfn2: function() {
                return this.x * 2;
            },
            x: 1,
            xfn4: function() {
                return this.x * 4;
            }
        };
        var data = random(tpl);
        data.x.should.eql(1);
        data.xfn2.should.eql(2);
        data.xfn4.should.eql(4);
    });

    it('holder', function() {
        rEmail.test(random('@EMAIL')).should.be.ok;
        rDate.test(random('@DATE')).should.be.ok;
        rTime.test(random('@TIME')).should.be.ok;
        rDatetime.test(random('@DATETIME')).should.be.ok;
    });

    it('complex', function() {
        var tpl = {
            'list|1-5': [{
                'id|+1': 1,
                'grade|1-100': 1,
                'float|1-100.1-10': 1,
                'star|1-5': '★',
                'published|1': false,
                'email': '@EMAIL',
                'date': '@DATE(HH:mm:ss)', // 属性 date 与 time 的格式互换
                'time': '@TIME(yyyy-MM-dd)',
                'datetime': '@DATETIME'
            }]
        };

        function validator(list) {
            var i, item, parts;
            (list.length >= 1 && list.length <= 5).should.be.ok;
            for (i = 0; i < list.length; i++) {
                item = list[i];

                (typeof item.id).should.eql('number');
                if (i > 0) item.id.should.eql(list[i - 1].id + 1);

                (typeof item.grade).should.eql('number');
                range(item.grade, 1, 100).should.be.ok;

                (typeof item.float).should.eql('number');
                range(item.float, 1, 101).should.be.ok;

                rFloat.lastIndex = 0;
                parts = rFloat.exec('' + item.float); // 可能会自动转为整数，例如 10.0 > 10
                if (parts[2]) {
                    (parts[2].length >= 1 && parts[2].length <= 10).should.be.ok;
                }

                range(item.star.length, 1, 5).should.be.ok;

                (typeof item.published).should.be.eql('boolean');

                rEmail.test(item.email).should.be.ok;
                rTime.test(item.date).should.be.ok;
                rDate.test(item.time).should.be.ok;
                rDatetime.test(item.datetime).should.be.ok;
            }
        }

        function target(tpl) {
            var list = random(tpl).list;
            validator(list);
        }

        for (var i = 0; i < 1000; i++) {
            target(tpl);
        }
    });

    it('escape', function() {
        var tpl;

        tpl = '\\@EMAIL';
        random(tpl).should.eql(tpl);

        tpl = '\\\\@EMAIL'; // TODO 如果有多个转义斜杠怎么办？只有奇数个转义斜杠，才会被认为是要转义 @
        random(tpl).should.eql(tpl);
    });

    it('reference', function() {
        var tpl = {
            name: '@first @last',
            first: 'fffffirst',
            last: 'lllllast'
        };
        var data = random(tpl);
        data.name.should.eql('fffffirst lllllast');
    });

    it('key order', function() {
        var tpl = {
            fn: function() {},
            first: '',
            second: '',
            third: ''
        };
        var data = random(tpl);
        var keys = _.keys(data);
        keys[0].should.eql('first');
        keys[1].should.eql('second');
        keys[2].should.eql('third');
        keys[3].should.eql('fn');
    });

    it('negative number', function() {
        var data;

        data = random({
            'number|-10--5': 1
        });
        range(data.number, -10, -5).should.be.ok;

        data = random({
            'number|-5--10': 1
        });
        range(data.number, -10, -5).should.be.ok;
    });
});


index.js
var Random = require('mock-random');
var _ = require('lodash').runInContext();

_.extend(_, {
    type: function (obj) {
        return (obj === null || obj === undefined) ? String(obj) : Object.prototype.toString.call(obj).match(/\[object (\w+)\]/)[1].toLowerCase();
    }
});

/*
    mock data
*/

/*
    rkey
        name|+inc
        name|repeat
        name|min-max
        name|min-max.dmin-dmax
        name|int.dmin-dmax

        1 name, 2 inc, 3 range, 4 decimal

    rplaceholder
        placeholder(*)

    [正则查看工具](http://www.regexper.com/)

    #26 生成规则 支持 负数，例如 number|-100-100
*/
var rkey = /(.+)\|(?:\+(\d+)|([\+\-]?\d+-?[\+\-]?\d*)?(?:\.(\d+-?\d*))?)/,
    rrange = /([\+\-]?\d+)-?([\+\-]?\d+)?/,
    rplaceholder = /\\*@([^@#%&()\?\s\/\.]+)(?:\((.*?)\))?/g; // (^(?:.|\r|\n)*?)

/*
    ## Mock

    ### Mock.mock( template|function() )

    根据数据模板生成模拟数据。
*/
module.exports = function(template) {
    return Handle.gen(template);
};

var Handle = {};

Handle.rule = function(name) {
    name = (name || '') + '';

    var parameters = (name || '').match(rkey),

        range = parameters && parameters[3] && parameters[3].match(rrange),
        min = range && parseInt(range[1], 10), // || 1
        max = range && parseInt(range[2], 10), // || 1
        // repeat || min-max || 1
        count = range ? !range[2] && parseInt(range[1], 10) || Random.integer(min, max) : 1,

        decimal = parameters && parameters[4] && parameters[4].match(rrange),
        dmin = decimal && parseInt(decimal[1], 10), // || 0,
        dmax = decimal && parseInt(decimal[2], 10), // || 0,
        // int || dmin-dmax || 0
        dcount = decimal ? !decimal[2] && parseInt(decimal[1], 10) || Random.integer(dmin, dmax) : 0,

        point = parameters && parameters[4];

    return {
        parameters: parameters,
        range: range,
        min: min,
        max: max,
        count: count,
        decimal: decimal,
        dmin: dmin,
        dmax: dmax,
        dcount: dcount,
        point: point
    };
};

/*
    template        属性值（即数据模板）
    name            属性名
    context         数据上下文，生成后的数据
    templateContext 模板上下文，

    Handle.gen(template, name, options)
    context
        currentContext, templateCurrentContext, 
        path, templatePath
        root, templateRoot
*/
Handle.gen = function(template, name, context) {
    name = name = (name || '') + '';

    context = context || {};
    context = {
        // 当前访问路径，只有属性名，不包括生成规则
        path: context.path || [],
        templatePath: context.templatePath || [],
        // 最终属性值的上下文
        currentContext: context.currentContext,
        // 属性值模板的上下文
        templateCurrentContext: context.templateCurrentContext || template,
        root: context.root,
        templateRoot: context.templateRoot
    };
    // console.log('path:', path.join('.'), template)

    var rule = Handle.rule(name);
    var type = _.type(template);

    if (Handle[type]) {
        return Handle[type]({
            // 属性值类型
            type: type,
            // 属性值模板
            template: template,
            // 属性名 + 生成规则
            name: name,
            // 属性名
            parsedName: name ? name.replace(rkey, '$1') : name,

            // 解析后的生成规则
            rule: rule,
            // 相关上下文
            context: context
        });
    }
    return template;
};

_.extend(Handle, {
    array: function(options) {
        var result = [], i;
        // 'arr': [{ 'email': '@EMAIL' }, { 'email': '@EMAIL' }]
        if (!options.rule.parameters) {
            for (i = 0; i < options.template.length; i++) {
                options.context.path.push(i);
                result.push(
                    Handle.gen(options.template[i], i, {
                        currentContext: result,
                        templateCurrentContext: options.template,
                        path: options.context.path
                    })
                );
                options.context.path.pop();
            }
        } else {
            // 'method|1': ['GET', 'POST', 'HEAD', 'DELETE']
            if (options.rule.min === 1 && isNaN(options.rule.max) && options.template.length > 1) {
                // fix #17
                options.context.path.push(options.name);
                result = Random.pick(Handle.gen(options.template, undefined, {
                    currentContext: result,
                    templateCurrentContext: options.template,
                    path: options.context.path
                }));
                options.context.path.pop();
            } else {
                // 'data|1-10': [{}, {}]
                for (i = 0; i < options.rule.count; i++) {
                    options.context.path.push(options.name);
                    result.push(Random.pick(Handle.gen(options.template, undefined, {
                        currentContext: result,
                        templateCurrentContext: options.template,
                        path: options.context.path
                    })));
                    options.context.path.pop();
                }
            }
        }
        return result;
    },
    object: function(options) {
        var result = {},
            keys, fnKeys, key, parsedKey, inc, i;

        // 'obj|min-max': {}
        if (options.rule.min) {
            keys = _.keys(options.template);
            keys = Random.shuffle(keys);
            keys = keys.slice(0, options.rule.count);
            for (i = 0; i < keys.length; i++) {
                key = keys[i];
                parsedKey = key.replace(rkey, '$1');
                options.context.path.push(parsedKey);
                result[parsedKey] = Handle.gen(options.template[key], key, {
                    currentContext: result,
                    templateCurrentContext: options.template,
                    path: options.context.path
                });
                options.context.path.pop();
            }

        } else {
            // 'obj': {}
            keys = [];
            fnKeys = []; // #25 改变了非函数属性的顺序，查找起来不方便
            for (key in options.template) {
                (typeof options.template[key] === 'function' ? fnKeys : keys).push(key);
            }
            keys = keys.concat(fnKeys);

            /*
            会改变非函数属性的顺序
            keys = _.keys(options.template)
            keys.sort(function(a, b) {
                var afn = typeof options.template[a] === 'function'
                var bfn = typeof options.template[b] === 'function'
                if (afn === bfn) return 0
                if (afn && !bfn) return 1
                if (!afn && bfn) return -1
            })
            */

            for (i = 0; i < keys.length; i++) {
                key = keys[i];
                parsedKey = key.replace(rkey, '$1');
                options.context.path.push(parsedKey);
                result[parsedKey] = Handle.gen(options.template[key], key, {
                    currentContext: result,
                    templateCurrentContext: options.template,
                    path: options.context.path
                });
                options.context.path.pop();
                // 'id|+1': 1
                inc = key.match(rkey);
                if (inc && inc[2] && _.type(options.template[key]) === 'number') {
                    options.template[key] += parseInt(inc[2], 10);
                }
            }
        }
        return result;
    },
    number: function(options) {
        var result, parts, i;
        if (options.rule.point) { // float
            options.template += '';
            parts = options.template.split('.');
            // 'float1|.1-10': 10,
            // 'float2|1-100.1-10': 1,
            // 'float3|999.1-10': 1,
            // 'float4|.3-10': 123.123,
            parts[0] = options.rule.range ? options.rule.count : parts[0];
            parts[1] = (parts[1] || '').slice(0, options.rule.dcount);
            for (i = 0; parts[1].length < options.rule.dcount; i++) {
                parts[1] += Random.character('number');
            }
            result = parseFloat(parts.join('.'), 10);
        } else { // integer
            // 'grade1|1-100': 1,
            result = options.rule.range && !options.rule.parameters[2] ? options.rule.count : options.template;
        }
        return result;
    },
    boolean: function(options) {
        var result;
        // 'prop|multiple': false, 当前值是相反值的概率倍数
        // 'prop|probability-probability': false, 当前值与相反值的概率
        result = options.rule.parameters ? Random.bool(options.rule.min, options.rule.max, options.template) : options.template;
        return result;
    },
    string: function(options) {
        var result = '',
            i, placeholders, ph, phed;
        if (options.template.length) {
            // 'star|1-5': '★',
            for (i = 0; i < options.rule.count; i++) {
                result += options.template;
            }
            // 'email|1-10': '@EMAIL, ',
            placeholders = result.match(rplaceholder) || []; // A-Z_0-9 > \w_
            for (i = 0; i < placeholders.length; i++) {
                ph = placeholders[i];
                // TODO 只有转义斜杠是偶数时，才不需要解析占位符？
                if (/^\\/.test(ph)) {
                    placeholders.splice(i--, 1);
                    continue;
                }
                phed = Handle.placeholder(ph, options.context.currentContext, options.context.templateCurrentContext);
                // 只有一个占位符，并且没有其他字符
                if (placeholders.length === 1 && ph === result && typeof phed !== typeof result) { // 
                    result = phed;
                    break;
                }
                result = result.replace(ph, phed);
            }
        } else {
            // 'ASCII|1-10': '',
            // 'ASCII': '',
            result = options.rule.range ? Random.string(options.rule.count) : options.template;
        }
        return result;
    },
    'function': function(options) {
        return options.template.call(options.context.currentContext);
    }
});

_.extend(Handle, {
    _all: function() {
        var re = {};
        for (var key in Random) re[key.toLowerCase()] = key;
        return re;
    },
    placeholder: function(placeholder, obj, templateContext) {
        // 1 key, 2 params
        rplaceholder.exec('');
        var parts = rplaceholder.exec(placeholder),
            key = parts && parts[1],
            lkey = key && key.toLowerCase(),
            okey = this._all()[lkey],
            params = parts && parts[2] || '';

        // 解析占位符的参数
        try {
            // 1. 尝试保持参数的类型
            /*
                #24 [Window Firefox 30.0 引用 占位符 抛错](https://github.com/nuysoft/Mock/issues/24)
                [BX9056: 各浏览器下 window.eval 方法的执行上下文存在差异](http://www.w3help.org/zh-cn/causes/BX9056)
                应该属于 Window Firefox 30.0 的 BUG
            */
            /* jshint -W061 */
            params = eval('(function(){ return [].splice.call(arguments, 0 ) })(' + params + ')');
        } catch (error) {
            // 2. 如果失败，只能解析为字符串
            // console.error(error)
            // if (error instanceof ReferenceError) params = parts[2].split(/,\s*/);
            // else throw error
            params = parts[2].split(/,\s*/);
        }

        // 占位符优先引用数据模板中的属性
        if (obj && (key in obj)) return obj[key];

        if (templateContext &&
            (typeof templateContext === 'object') &&
            (key in templateContext) &&
            (placeholder !== templateContext[key]) // fix #15 避免自己依赖自己
        ) {
            templateContext[key] = Handle.gen(templateContext[key], key, {
                currentContext: obj,
                templateCurrentContext: templateContext
            });
            return templateContext[key];
        }

        if (!(key in Random) && !(lkey in Random) && !(okey in Random)) return placeholder;

        for (var i = 0; i < params.length; i++) {
            rplaceholder.exec('');
            if (rplaceholder.test(params[i])) {
                params[i] = Handle.placeholder(params[i], obj);
            }
        }

        var handle = Random[key] || Random[lkey] || Random[okey];
        switch (_.type(handle)) {
            case 'array':
                return Random.pick(handle);
            case 'function':
                var re = handle.apply(Random, params);
                if (re === undefined) re = '';
                return re;
        }
    }
});
